### ISSUE TITLE
master often hangs on the "jobs running tty-test program resizing window" functional test

### ISSUE THREAD

**Issue by jszakmeister** [neovim/neovim#2225]
I noticed the FreeBSD and OS X build agents hanging on functional tests.  I can reproduce it my normal working tree, and it's hanging on the "jobs running tty-test program resizing window" test.  I'm not quite sure why yet, but it never recovers (one agent was running for nearly 7 hours before I killed it and put timeouts on the test).  I also tried using #2221 to see if it would help, but it didn't. :-(

**Comment by tarruda** [comment]
Does the problem persists if you apply [this commit](https://github.com/tarruda/neovim/commit/a11c1d0cc5b48e169d3761d4a0673ce38b34709f)?

**Comment by jszakmeister** [comment]
Unfortunately, it does persist. :-(

**Comment by tarruda** [comment]
What is the output when you apply this patch and only run that specific test?

```diff
diff --git a/test/functional/job/job_spec.lua b/test/functional/job/job_spec.lua
index 0d561ec..f84878f 100644
--- a/test/functional/job/job_spec.lua
+++ b/test/functional/job/job_spec.lua
@@ -142,6 +142,7 @@ describe('jobs', function()
       local rv = ''
       while true do
         local msg = next_message()
+        io.stderr:write(require('inspect')(msg) .. '\n'); io.stderr:flush()
         local data = msg[3][1]
         for i = 1, #data do
           data[i] = data[i]:gsub('\n', '\000')
```

Note that it requires the 'inspect' lua package, which you can install with `./.deps/usr/bin/luarocks install inspect`

**Comment by jszakmeister** [comment]
How do I run just that test?  I run the whole file, but I'm not sure how to run just a single test.

**Comment by jszakmeister** [comment]
Also, I'm not sure if it reproduces when running `job_spec.lua` by itself--I haven't tried that.  One quirk I have noticed: it reproduces better on a clean build.  My time this morning is running out...

**Comment by jszakmeister** [comment]
Here's the result of running the whole file:

```text
./test/functional/job/job_spec.lua:
  jobs returns 0 when it fails to start ... ●
  jobs calls JobActivity when the job writes and exits ... ●
  jobs allows interactive commands ... ●
  jobs preserves NULs ... ●
  jobs will not buffer data if it doesnt end in newlines ... ●
  jobs can preserve newlines ... ●
  jobs can preserve nuls ... ●
  jobs can avoid sending final newline ... ●
  jobs will not allow jobsend/stop on a non-existent job ... ●
  jobs will not allow jobstop twice on the same job ... ●
  jobs will not cause a memory leak if we leave a job running ... ●
  jobs will only emit the "exit" event after "stdout" and "stderr" ... ●
  jobs running tty-test program echoing input ... { "notification", "stdout", { { "tty ready\r", "" } } }
{ "notification", "stdout", { { "test" } } }
●
  jobs running tty-test program resizing window ... { "notification", "stdout", { { "tty ready\r", "" } } }
{ "notification", "stdout", { { "screen resized. rows: 10, columns: 40\r", "" } } }
{ "notification", "stdout", { { "screen resized. rows: 40, columns: 10\r", "" } } }
●
  jobs running tty-test program preprocessing ctrl+c with terminal driver ... ●
13 successes / 0 failures / 0 errors / 2 pending : 0.0 seconds

Pending → ./test/functional/job/job_spec.lua @ 186
jobs running tty-test program preprocessing ctrl+c with terminal driver

Pending → ./test/functional/job/job_spec.lua @ 130
jobs will only emit the "exit" event after "stdout" and "stderr"
```

**Comment by jszakmeister** [comment]
It appears simply having that inspect line allows the test to move on. :-(

**Comment by tarruda** [comment]
> It appears simply having that inspect line allows the test to move on. :-(

Certainly a race condition then, I will inspect the test a little closer. I think its time for me to build a BSD virtual machine

**Comment by tarruda** [comment]
Ok, this one might still reproduce the problem:

```diff
diff --git a/test/functional/job/job_spec.lua b/test/functional/job/job_spec.lua
index 0d561ec..c2d53f0 100644
--- a/test/functional/job/job_spec.lua
+++ b/test/functional/job/job_spec.lua
@@ -175,10 +175,15 @@ describe('jobs', function()
     end)

     it('resizing window', function()
+      io.stderr:write('0\n'); io.stderr:flush()
       nvim('command', 'call jobresize(j, 40, 10)')
+      io.stderr:write('1\n'); io.stderr:flush()
       eq('screen resized. rows: 10, columns: 40', next_chunk())
+      io.stderr:write('2\n'); io.stderr:flush()
       nvim('command', 'call jobresize(j, 10, 40)')
+      io.stderr:write('3\n'); io.stderr:flush()
       eq('screen resized. rows: 40, columns: 10', next_chunk())
+      io.stderr:write('4\n'); io.stderr:flush()
     end)

     -- FIXME This test is flawed because there is no telling when the OS will send chunks of data.
```

**Comment by jszakmeister** [comment]
Yes, that hangs:

```text
./test/functional/job/job_spec.lua:
  jobs returns 0 when it fails to start ... ●
  jobs calls JobActivity when the job writes and exits ... ●
  jobs allows interactive commands ... ●
  jobs preserves NULs ... ●
  jobs will not buffer data if it doesnt end in newlines ... ●
  jobs can preserve newlines ... ●
  jobs can preserve nuls ... ●
  jobs can avoid sending final newline ... ●
  jobs will not allow jobsend/stop on a non-existent job ... ●
  jobs will not allow jobstop twice on the same job ... ●
  jobs will not cause a memory leak if we leave a job running ... ●
  jobs will only emit the "exit" event after "stdout" and "stderr" ... ●
  jobs running tty-test program echoing input ... ●
  jobs running tty-test program resizing window ... 0
1
```

**Comment by tarruda** [comment]
Can you reproduce the problem with this patch?

```diff
diff --git a/test/functional/job/job_spec.lua b/test/functional/job/job_spec.lua
index 0d561ec..1416499 100644
--- a/test/functional/job/job_spec.lua
+++ b/test/functional/job/job_spec.lua
@@ -175,6 +175,7 @@ describe('jobs', function()
     end)

     it('resizing window', function()
+      nvim('command', 'sleep 1')
       nvim('command', 'call jobresize(j, 40, 10)')
       eq('screen resized. rows: 10, columns: 40', next_chunk())
       nvim('command', 'call jobresize(j, 10, 40)')
```

**Comment by jszakmeister** [comment]
I've tried several times, and with the sleep, it doesn't reproduce.

**Comment by tarruda** [comment]
It seems like the problem lies in the tty-test program, which is not receiving `SIGWINCH` when sent too soon. My guess is that `SIGWINCH` is being received before libuv installs the signal handler, and if so this patch might fix the problem:

```diff
diff --git a/test/functional/job/tty-test.c b/test/functional/job/tty-test.c
index 25e7684..b395c9c 100644
--- a/test/functional/job/tty-test.c
+++ b/test/functional/job/tty-test.c
@@ -80,6 +80,12 @@ static void read_cb(uv_stream_t *stream, ssize_t cnt, const uv_buf_t *buf)
   free(buf->base);
 }

+static void prepare_cb(uv_prepare_t *handle)
+{
+  fprintf(stderr, "tty ready\n");
+  uv_prepare_stop(handle);
+}
+
 int main(int argc, char **argv)
 {
   if (!is_terminal(stdin)) {
@@ -98,7 +104,9 @@ int main(int argc, char **argv)
   }

   bool interrupted = false;
-  fprintf(stderr, "tty ready\n");
+  uv_prepare_t prepare;
+  uv_prepare_init(uv_default_loop(), &prepare);
+  uv_prepare_start(&prepare, prepare_cb);
   uv_tty_t tty;
   uv_tty_init(uv_default_loop(), &tty, fileno(stderr), 1);
   uv_read_start((uv_stream_t *)&tty, alloc_cb, read_cb);
```

**Comment by jszakmeister** [comment]
@tarruda This last patch seems to be doing the trick.  Do you think we can get that on master some time soon?

**Comment by tarruda** [comment]
> @tarruda This last patch seems to be doing the trick. Do you think we can get that on master some time soon? 

In a few minutes I will prepare a PR with it and a few extra fixes currently in #2076

**Comment by jszakmeister** [comment]
Awesome.  Thanks a lot for helping to work through the problem!
